build.sc

```diff
@@ -1631,6 +1633,134 @@ object dev extends MillModule {
 
 /** Generates the mill documentation with Antora. */
 object docs extends Module {
+  // This module isn't really a ScalaModule, but we use it to generate
+  // consolidated documentation using the Scaladoc tool.
+  object site extends ScaladocSiteModule {
+    def scalaVersion = Deps.scalaVersion
+
+    // specify subdirectory for scaladoc
+    def scaladocSubPath: os.SubPath = os.sub / "api" / "latest"
+
+    def moduleDeps = build.millInternal.modules.collect { case m: MillApiModule => m}
+
+    def scaladocPushGitURI = "git@github.com:com-lihaoyi/mill.git"
+  }
+
+  // This module isn't really a ScalaModule, but we use it to generate
+  // consolidated documentation using the Scaladoc tool.
+  trait ScaladocSiteModule extends ScalaModule {
+    def moduleDeps: Seq[JavaModule]
+
+    override def compileClasspath = T{
+      super.compileClasspath() ++
+      T.traverse(moduleDeps)(_.compileClasspath)().flatten
+    }
+    // stage the static website and/or doc into the 'stage' task destination
+    // directory adapted from: https://github.com/com-lihaoyi/mill/discussions/1194
+    def stage = T {
+      import mill.eval.Result
+
+      if (!os.isDir(millSourcePath)) {
+        T.log.info(s"""Source path "${millSourcePath}" not found, ignoring""")
+        T.log.info(s"Staging index.html from method defaultSiteIndex")
+        os.write.over(T.dest / "index.html", defaultSiteIndex)
+      } else {
+        val sitefiles = os.walk(millSourcePath, skip = (p: os.Path) => !os.isFile(p))
+        T.log.info(s"Staging ${sitefiles.length} site files from site source path ${millSourcePath}")
+        for (f <- sitefiles ){
+          os.copy.over(f, T.dest / f.subRelativeTo(millSourcePath), createFolders = true)
+        }
+      }
+
+      val scaladocFiles: Seq[os.Path] =
+        T.traverse(moduleDeps)(_.allSourceFiles)().flatten.map(_.path)
+
+      T.log.info(s"Staging scaladoc for ${scaladocFiles.length} files")
+
+      val scaladocPath: os.Path = T.dest / scaladocSubPath
+      os.makeDir.all(scaladocPath)
+
+      // the details of the options and zincWorker call are significantly
+      // different between scala-2 scaladoc and scala-3 scaladoc
+      // below is for scala-2 variant
+      val options: Seq[String] = Seq(
+        "-doc-title", "Mill",
+        "-doc-version", millVersion(),
+        "-d", scaladocPath.toString,
+        "-classpath", compileClasspath().map(_.path).mkString(":"),
+      )
+
+      val docReturn = zincWorker.worker().docJar(
+        scalaVersion(),
+        scalaOrganization(),
+        scalaDocClasspath().map(_.path),
+        scalacPluginClasspath().map(_.path),
+        options ++ scaladocFiles.map(_.toString)
+      ) match {
+        case true => Result.Success(PathRef(T.dest))
+        case false => Result.Failure("doc generation failed")
+      }
+
+      docReturn
+    }
+
+    def push() = T.command {
```

Comment:
Instead of manual pushing to some site, we could publish a synthetic artifact and upload the scaladoc with the javadoc classifier to Sonatype. There are already lots of nice online viewers like https://javadoc.io/doc/com.lihaoyi/mill-scalalib_2.13/latest/index.html

Reply:
The `def push()` stuff was copy-pasta from the original source, but I was thinking we could just upload the scaladoc HTML to github pages together with the Antora docsite, since we already have to deploy that stuff anyway